plugins/rts54hub/fu-rts54hub-rtd21xx-device.c

```diff
@@ -160,6 +207,56 @@ fu_rts54hub_rtd21xx_device_i2c_read(FuRts54hubRtd21xxDevice *self,
 	return TRUE;
 }
 
+gboolean
+fu_rts54hub_rtd21xx_device_ddcci_read(FuRts54hubRtd21xxDevice *self,
+				      guint8 target_addr,
+				      guint8 sub_addr,
+				      guint8 *data,
+				      gsize datasz,
+				      GError **error)
+{
+	guint8 checksum = 0x50;
+	guint8 buf_read[MAX_READ_WRITE_LENGTH_ONE_TIME] = {0x00};
+	gsize i = 0;
+
+	if (datasz > MAX_READ_WRITE_LENGTH_ONE_TIME) {
+		g_prefix_error(error,
+			       "ddcci read length exceed max length:%i: ",
+			       MAX_READ_WRITE_LENGTH_ONE_TIME);
+		return FALSE;
+	}
+
+	if (!fu_rts54hub_rtd21xx_device_i2c_read(self,
+						 target_addr,
+						 sub_addr,
+						 buf_read,
+						 datasz,
+						 error)) {
+		g_prefix_error(error, "failed to DDCCI read I2C: ");
+		return FALSE;
+	}
+
+	if (buf_read[0] != target_addr) {
+		g_prefix_error(error, "failed to DDCCI read I2C target addr invalid: ");
+		return FALSE;
+	}
+
+	for (i = 0; i < (gsize)((buf_read[1] & 0x7F) + 2); i++) {
+		checksum ^= buf_read[i];
+	}
+
+	if (checksum != buf_read[(buf_read[1] & 0x7F) + 2]) {
+		g_prefix_error(error, "failed to DDCCI read I2C checksum error: ");
+		return FALSE;
+	}
+
+	for (i = 0; i < (gsize)(buf_read[1] & 0x7f) + 3; i++) {
```

Comment:
this needs an explicit check against datasz; it's an untrusted value and can easily overflow data[]